### DISCUSSION TITLE
How to use a URL parameter to set the default value of a radio button (Next.js + react-hook-form)

### DISCUSSION THREAD

**Discussion by ben519** [react-hook-form/react-hook-form#9887] [Q&A] [answered]
Cross-posting this from [StackOverflow](https://stackoverflow.com/questions/75376534/how-to-use-url-parameter-to-set-the-default-value-of-a-radio-button).

## The Goal

I'm building a two-sided marketplace for buyers and sellers. When someone navigates to my signup page, they could hit a URL like one of the following

- `/signup`
- `/signup?accountType=buyer`
- `/signup?accountType=seller`
- `/signup?accountType=asdfghjkl` (nonsensical, but possible)

My signup page has a radio button input where they choose Buyer or Seller.

<img width="446" alt="Screenshot 2023-02-07 at 10 20 11 AM" src="https://user-images.githubusercontent.com/723099/217347713-592a1b77-20fd-4572-950c-e00b171cde1a.png">

**Rules** 

- The user must select one of these options.
- If the URL contains `accountType=buyer`, I would like to set the *default* selection as Buyer
- If the URL contains `accountType=seller`, I would like to set the *default* selection as Seller
- Even if a default option is selected, the user should be able to change it

----

## What I've tried

I am struggling to make this work with Next.js and [react-hook-form](https://react-hook-form.com/). Here's what I've tried.

```
// Fields.jsx

import { forwardRef } from 'react'

function Label({ id, children }) {
  return (
    <label htmlFor={id}>
      {children}
    </label>
  )
}

export const RadioFieldWithRef = forwardRef(function RadioField({ id, label, options, name, className = '', ...props }, ref) {
  return (
    <div className={className}>
      {label && <Label id={id}>{label}</Label>}
      <div>
        {options.map((option) => (
          <div className="flex items-center" key={option.value}>
            <input
              id={option.value}
              name={name}
              type="radio"
              value={option.value}
              defaultChecked={option.defaultChecked}
              ref={ref}
              {...props}
            />
            <label htmlFor={option.value}>
              {option.label}
            </label>
          </div>
        ))}
      </div>
    </div>
  )
})
```

```
// signup.jsx

import { useRouter } from 'next/router'
import { RadioFieldWithRef } from '@/components/Fields'
import { useForm } from "react-hook-form";

export default function Signup() {
  const router = useRouter()
  const { accountType } = router.query
  
  // Email & Password Sign Up Form
  const { register } = useForm();
  const accountTypeField = register("account_type", {
    required: "Must select account type"
  })

  return (
        <form>
          <RadioFieldWithRef
            label="I'm a ..."
            name="account_type"
            options={ [
              {
                label: 'Buyer',
                value: 'buyer',
                defaultChecked: accountType === "buyer"
              },
              {
                label: 'Seller',
                value: 'seller',
                defaultChecked: accountType === "seller"
              },
            ] }
            {...accountTypeField}
          />
            <button type="submit">Submit</button>
        </form>
  )
}
```

### The Problem
When I try a URL like `/signup?accountType=buyer`, the default selection is not being set. I think this is because `router.query` is actually undefined on the first render. `console.log("accountType", accountType)` shows `undefined` before it eventually shows `buyer`. But I'm not sure how to overcome this.

(I've also tried using a [Controller](https://react-hook-form.com/api/usecontroller/controller), but couldn't figure out how to update `field.value` with the value taken from the URL parameter. I started to set up a [reprex here](https://codesandbox.io/s/awesome-elbakyan-n1dx05?file=/src/App.js), but was unable to mimic URL parameters.

  [1]: https://i.stack.imgur.com/y4cpKl.png

**Comment by leapful** [comment, marked as answer]
@ben519 

You need to watch on change of `router.query` by using `useEffect` hook then use `setValue()` method from `useForm()` hook to set  value for field `account_type`.

Working demo: https://codesandbox.io/s/divine-hooks-7yb3bm?file=/src/App.js

**Reply by ben519** [reply]
Ah, this makes sense. Thanks so much!
